### ISSUE TITLE
Variable bandwidth for 3 dimensional data.

### ISSUE THREAD

**Issue by ytarricq** [tommyod/KDEpy#26]
Hello, 

First of all, thanks for the great package.
I'm trying to compute density maps of a 3 dimensional points distribution. I understood from the documentation that a variable bandwith method was available but I couldn't figure out how to set up this option. 
Additionnaly, in the case of a fixed bandwidth KDE for multidimensional data, I would have expected as in the stats_models_multivariateKDE implementation to be able to use a bandwidth per dimension but it seems that we can either use a single value of the bandwidth or to use one bandwidth per data point. Is it in order to take into account the weight of each data point that you implemented it this way ?

Thanks in advance.

Cheers
Yoann

**Comment by tommyod** [comment]
Thanks for the kind words, and for raising this issue @ytarricq .

- **Variable bandwidth** (i.e. a unique bandwidth per data point) is only available in the NaiveKDE and TreeKDE implementations. You have to supply an array as the `bw` parameter, see [the docs here](https://kdepy.readthedocs.io/en/latest/API.html#naivekde).
- If you want to use a **bandwidth matrix that depends on the dimensions**, i.e. bandwidth 2 in the `x` direction and bandwidth 3 in the `y` dimension, that's not supported directly. The reason is that every kernel is implement as a [radial basis function](https://en.wikipedia.org/wiki/Radial_basis_function). Scipy supports arbitrary bandwidth matrices, which is easy for Gaussian kernels. KDEpy supports arbitrary kernels, which makes this tougher. There's an elegant workaround to this problem: use the SVD to transform the data (scale, rotate) instead, see [my recipe here](https://github.com/tommyod/KDEpy/issues/6#issuecomment-446711581).

Hope this helps you. :+1:

Making that recipe idiot-proof and implementing it in the main library would be a good task. If you (or anyone reading this) is up for it, that's a PR I would merge.

**Comment by ytarricq** [comment]
Thanks for the quick answer ! Made things clearer between the bandwidth matrices/variable bandwidths. 
I will work on the best way to handle my data and will get back at you if I'm successfull.

**Comment by philippeller** [comment]
I'm having the same issue with the fixed bandwidth for all dimensions.
In my case one dimension has a radically different scale than the others and hence the resulting KDEs don't look good. Just scaling the data in that dimension works fine (and rescaling after KDE), since I don't need any rotation/covariance.
Wouldn't implementing a bw per dimensions, if given as an iterable, not get us a long way without complicating things too much?

**Comment by tommyod** [comment]
@philippeller : Since the kernel functions are radial basis functions, since would amount to scaling the data on input in each dimensions, computing the KDE, then scaling back. However, it would hide *how* the data is scaled to the user. Some options are: min/max scaling, standardizing with the standard deviation and the mean, quantile transformations, etc.

I feel that "*simple is better than complex*" and the "*principle of least surprise*" applies here. Doing some implicit scaling scheme might confuse users. Stating that "the multidimensional KDE is anisotropic" and letting users handle scaling seems simpler to understand and less likely to produce unexpected results. 

I'm open to suggestions of course. But I would need some details.

**Comment by philippeller** [comment]
Maybe I'm missing some important point, but I was thinking not an implicit, but an explicit scaling.

Let's say the user supplies two dimensional data (`x` and `y`) and the fixed bandwidths as `(bw_x, bw_y`).
Now internally you compute `scaled_x` = `x` / `bw_x`( and `scaled_y` = `x` / `bw_y`), then proceed with the KDE on the scaled data using bandwidth = 1, and in the end just undo the scaling, wouldn't that work?

**Comment by tommyod** [comment]
That would work. :+1:  Thanks for clarifying. I got a little ahead of myself.

What you're sketching might be worth implementing. In a different issue #6 we had some discussions about a more general case. It's really an issue of API design. The way I see it:

Pros:
- Useful in some use cases, saves the users some time (but they can do it themselves too)

Cons:
- Extends the current API a little (but backwards compatible, so no big deal)
- Doesn't implement the more general case (general anisotropic KDEs via rotations)

**In conclusion I would merge a PR that implements this.** :+1:  No promises about when/if I'll find time to do it myself though.
